### ISSUE TITLE
Clarification: allow MPI_PROC_NULL as neighbor in (distributed) graph topologies

### ISSUE THREAD

**Issue by omor1** [mpi-forum/mpi-issues#87]
This was originally posted on the mpi-comments mailing list, but as @AndrewGaspar noted in #86, this repository is much more active.

# Problem

The topology creation functions (`MPI_GRAPH_CREATE`, `MPI_DIST_GRAPH_CREATE`, and `MPI_DIST_GRAPH_CREATE_ADJACENT`) define a virtual topology, described as a communication graph. Edges are defined between sources and destinations, both of which are ranks of processes in the input communicator [0].
In some instances, such as a communication graph where most nodes have the same number of neighbors (for instance a binary tree topology), it may be useful to have a "dummy" source or destination. This is provided for in MPI as `MPI_PROC_NULL` [1]. Then a caller can have boundary nodes (such as a root or leaves in a tree) communicate with these "dummy" neighbors, simplifying topology code logic.
This in analogous to what is done in non-periodic cartesian topologies [2]; MPI 3.1 documentation on `MPI_CART_SHIFT` [3] states:

>Depending on the periodicity of the Cartesian group in the specified coordinate direction, `MPI_CART_SHIFT` provides the identifiers for a circular or an end-off shift. In the case of an end-off shift, the value `MPI_PROC_NULL` may be returned in `rank_source` or `rank_dest`, indicating that the source or the destination for the shift is out of range.

Additionally, the MPI 3.1 documentation on Neighborhood Collective Communication on Process Topologies [4] states:

>For a Cartesian topology, created with `MPI_Cart_create`, the sequence of neighbors in the send and receive buffers at each process is defined by order of the dimensions, first the neighbor in the negative direction and then in the positive direction with displacement 1. The numbers of sources and destinations in the communication routines are `2*ndims` with `ndims` defined in `MPI_Cart_create`. If a neighbor does not exist, i.e., at the border of a Cartesian topology in the case of a non-periodic virtual grid dimension (i.e., `periods[ . . . ]==false`), then this neighbor is defined to be `MPI_PROC_NULL`.
>
>If a neighbor in any of the functions is `MPI_PROC_NULL`, then the neighborhood collective communication behaves like a point-to-point communication with `MPI_PROC_NULL` in this direction. That is, the buffer is still part of the sequence of neighbors but it is neither communicated nor updated.

There is some additional discussion with @bosilca in [5].

# Proposal

Explicitly allow `MPI_PROC_NULL` as a neighbor in graph topologies. Communication with such a neighbor in the neighborhood communication collectives (`MPI_(I)NEIGHBOR_ALLGATHER(,V)`, `MPI_(I)NEIGHBOR_ALLTOALL(,V,W)`) is defined identically to point-to-point communications with `MPI_PROC_NULL` [1]:

>A communication with process `MPI_PROC_NULL` has no effect. A send to `MPI_PROC_NULL` succeeds and returns as soon as possible. A receive from `MPI_PROC_NULL` succeeds and returns as soon as possible with no modifications to the receive buffer.

This is, as stated in [4], indeed already in the standard, and merely needs a clarification that `MPI_PROC_NULL` is *explicitly* permitted as a neighbor in graph topologies.

# Changes to the Text

On pages 295, 297, and 299, clarify that `MPI_PROC_NULL` may be the neighbor of a process. Sample wording includes:

>Processes with `MPI_PROC_NULL` neighbors are allowed.
> `MPI_PROC_NULL` may be a neighbor of a process.

# Impact on Implementations

Implementations will be required to permit `MPI_PROC_NULL` as a valid neighbor in graph topologies, if they do not already. In most cases, this should be fairly simple: see [6].

# Impact on Users

None for current users; all existing code will remain valid. This may open up new possibilities and allow for some simplification of existing code.

# References

[0]: MPI 3.1, Section 7.5.3 & 7.5.4, Pages 294–302
[1]: MPI 3.1, Section 3.11, Pages 80–81
[2]: MPI 3.1, Section 7.5.1, Page 292
[3]: MPI 3.1, Section 7.5.6, Page 310
[4]: MPI 3.1, Section 7.5, Pages 314–315
[5] open-mpi/ompi#4675
[6] open-mpi/ompi#4898

**Comment by hjelmn** [comment]
I don't see the benefit of this change. Can you give an example of what would be easier if MPI_PROC_NULL is allowed for dist graph?

**Comment by omor1** [comment]
A clear example is a virtual binary tree topology. Nodes have between 1 and 3 neighbors (a parent and up to two children); most nodes are inner nodes and all three, some a leaves and have one, and a few have two (the root and some inner nodes in a non-full tree).

Currently, this must be implemented as follows:

```C
MPI_Comm tree;
int world_rank;
int world_size;
int neighbors[3];
int n = 0;

MPI_Comm_rank(MPI_COMM_WORLD, &world_rank);
MPI_Comm_size(MPI_COMM_WORLD, &world_size);

if (world_rank > 0) 
        neighbors[n++] = (world_rank-1)/2;
if (2*world_rank+1 < world_size)
        neighbors[n++] = 2*world_rank+1;
if (2*world_rank+2 < world_size)
        neighbors[n++] = 2*world_rank+2;

MPI_Dist_graph_create_adjacent(MPI_COMM_WORLD,
                               n, neighbors, MPI_UNWEIGHTED,
                               n, neighbors, MPI_UNWEIGHTED,
                               MPI_INFO_NULL, true, &tree_comm);
```

It then becomes difficult to tell what is in `neighbors[0]`; is this a parent, or a left child? This differs depending on whether or not you are the root. The same is true for `neighbors[1]`; this could be a left child, a right child, or a bogus value.

Suppose I later want to have every node send something to their left child; to do so, I must either store the left child separately from the neighbors array, or store what index in the neighbors array contains the left child, or recalculate either of those on demand—each time!

While this isn't difficult for a binary tree, it is not difficult to see how in a more general topology that is mostly homogeneous this may become a burden.

With this clarification, this instead becomes

```C
MPI_Comm tree;
int world_rank;
int world_size;
int neighbors[3];

MPI_Comm_rank(MPI_COMM_WORLD, &world_rank);
MPI_Comm_size(MPI_COMM_WORLD, &world_size);

neighbors[0] = world_rank > 0              ? (world_rank-1)/2 : MPI_PROC_NULL;
neighbors[1] = 2*world_rank+1 < world_size ? 2*world_rank+1   : MPI_PROC_NULL;
neighbors[2] = 2*world_rank+2 < world_size ? 2*world_rank+2   : MPI_PROC_NULL;

MPI_Dist_graph_create_adjacent(MPI_COMM_WORLD,
                               3, neighbors, MPI_UNWEIGHTED,
                               3, neighbors, MPI_UNWEIGHTED,
                               MPI_INFO_NULL, true, &CommTree);
```

Now `neighbors[0]` is _always_ the parent; `neighbors[1]` is _always_ the left child; and `neighbors[2]` is always the right child. These may be nonexistent (`MPI_PROC_NULL`), but sending and receiving from them is never incorrect. In short, this helps simplify operations near the boundaries—exactly what `MPI_PROC_NULL` is meant to be used for!

>In many instances, it is convenient to specify a "dummy" source or destination for communication. This simplifies the code that is needed for dealing with boundaries, for example, in the case of a non-circular shift done with calls to send-receive. (MPI 3.1, Section 3.11, Page 80)

A further point is that currently, a distributed graph topology _cannot emulate a cartesian topology_! I believe that this is not the intention of the MPI specification; a graph topology is intended to be the most general, and it is only because cartesian topologies are so common that the cartesian topology functions are provided (so as to simplify the process).

**Comment by omor1** [comment]
Also note that MPI states:

>The special value MPI_PROC_NULL can be used instead of a rank wherever a source or a destination argument is required in a call.

So that a strict reading of the text should require this anyway; this proposal merely seeks to clarify the fact, as some implementations do not currently support this.

(Actually, I've gone and tested—it seems that while OpenMPI's behavior is erroneous, MPICH's appears to be correct. In any case, a clarification may be useful.)

**Comment by omor1** [comment]
Given that `MPI_PROC_NULL` is supposed to be able to be used wherever a rank is required as a source or destination, and coupled with the fact that the neighborhood collectives *explicitly* support `MPI_PROC_NULL` neighbors, rejecting a topology in which `MPI_PROC_NULL` is specified as the rank of a neighbor would appear to be erroneous—at least to me.

What is your reasoning for the validity of Open MPI's behavior? Keep in mind that the neighbors at the borders of non-periodic dimensions of Cartesian topologies are explicitly defined to be `MPI_PROC_NULL` by the standard.

It is interesting to note that, as far as I can tell, the distributed graph topology constructors are the only place in the standard where ranks are specified to be 'non-negative integers'; all other functions merely specify them as 'integers'. Even the older non-distributed graph topology constructor specifies the flattened edge representation of neighbors as 'integers'.

**Comment by hjelmn** [comment]
Exactly, the sources and destinations are *non-negative*. This requirement forbids MPI_PROC_NULL which by definition MUST be negative to avoid conflicting with another valid rank. As such Open MPI's behavior is not erroneous. To me it is also logical.

This behavior may or may not have been the original intent and there may be need for an errata (at which point we will change the behavior of Open MPI). This will certainly be discussed at the MPI Forum meeting in June.

**Comment by jdinan** [comment]
Does MPI require implementations to support MAX_INT processes?  I'm not sure that _non-negative_ in this context was intended to exclude MPI_PROC_NULL.

**Comment by tonyskjellum** [comment]
I have to agree — MPI_PROC_NULL has no formal value —- it is special 

Anthony Skjellum, PhD
205-807-4968

**Comment by omor1** [comment]
As far as I can tell, there is no requirement that an MPI implementation support `INT_MAX` processes or that `MPI_PROC_NULL` be negative. It _is_ usually implemented as a negative integer (`-2` in Open MPI, `-1` in MPICH), but this is not required—it _could_ conceivably be implemented as `INT_MAX` in an implementation that supported less than `INT_MAX` processes.

According to @hjelmn's reasoning, that would make whether or not `MPI_PROC_NULL` is a valid neighbor an artifact of the implementation, which is probably not the intent.

**Comment by hjelmn** [comment]
FWIW. My interpretation is based on there being no material benefit to allowing null. It saves some minimal bookkeeping for apps. I can see having a limitation allowing for some optimization in the library. There is a cost to allowing null.

**Comment by omor1** [comment]
Distributed graph topologies were added in MPI-2.2. The author for Process Topologies for MPI-2.2 was Torsten Hoefler. He was also the editor and organizer for that chapter for MPI-3.0 and MPI-3.1 and is still the chair for the upcoming MPI-4.0.

Certainly, every feature has a cost associated with it—hence why determining what the correct behavior ought to be and ensuring that implementations function according to said behavior is important.

**Comment by wgropp** [comment]
Also, we went through a similar case for RMA - some implementations rejected MPI_PROC_NULL in communication calls such as MPI_Put.  We viewed this as an errata and eventually added a specific statement to this effect; see 11.3

> MPI_PROC_NULL is a valid target rank in all MPI RMA communication calls.  
> 
This was a clarification because of this text from section 3.11:

> The special value MPI_PROC_NULL can be used instead of a rank wherever a source or a destination argument is required in a call.  
> 

This is in the context of “communication” but applies to all routines.

Yes, there is an overhead to checking for this. That is far more serious for the communication calls, particularly the RMA ones, so it is clear that the Forum has already decided that the extra overhead of checking for MPI_PROC_NULL is not a factor.

Bill

William Gropp
Director and Chief Scientist, NCSA
Thomas M. Siebel Chair in Computer Science
University of Illinois Urbana-Champaign

**Comment by hjelmn** [comment]
@wgropp Ok, then we probably should fix this by dropping non-negative so that the argument descriptions match the other calls. Is this a ticket 0 change?

**Comment by dholmes-epcc-ed-ac-uk** [comment]
I think there is a reasonable argument for choosing to make this an errata in order to increase its visibility. I have always assumed, through osmosis, that MPI_PROC_NULL is not permitted for (dist) graph topologies. I agree that there is no strong indication in the text of the MPI Standard stating that restriction but there is a clear general statement to the contrary. This may be a common mis-conception and it may behove the MPI Forum to publicise this change much more loudly than is typically done for a ticket 0 change.

@wgropp so, for dist graph, the text for the sources and destinations arguments would read "array of valid rank values" instead of "array of non-negative integers"? And, for graph, the text for the edges argument would read "array of valid rank values describing graph edges (see below)"?

I think an explicit sentence akin to the RMA errata should be added for each case as well.
For dist graph, how about adding (at line 34 on page 297):
> MPI_PROC_NULL is a valid rank value for \mpiarg{sources} or for \mpiarg{destinations}.

For graph, how about adding (at line 43 on page 294):
> MPI_PROC_NULL is a valid rank value for \mpiarg{edges}.

**Comment by wgropp** [comment]
I agree with this.  While the standard document is not a user manual, clarity is important, and these changes significantly increase the clarity without adding much text.

Bill

William Gropp
Director and Chief Scientist, NCSA
Thomas M. Siebel Chair in Computer Science
University of Illinois Urbana-Champaign

**Comment by dholmes-epcc-ed-ac-uk** [comment]
Suggested change, based on MPI-3.x (as of a few moments ago), with change-bars and change log entry:
[mpi-report-ticket87.pdf](https://github.com/mpi-forum/mpi-standard/files/2026783/mpi-report-ticket87.pdf)

**Comment by dholmes-epcc-ed-ac-uk** [comment]
Suggested change, based on MPI-3.x (as of a few moments ago), with change-bars and change log entry:
[mpi-report-ticket87.pdf](https://github.com/mpi-forum/mpi-issues/files/2027399/mpi-report-ticket87.pdf)

**Comment by omor1** [comment]
1. "array of non-negative integers" should be replaced by "array of valid rank values" in `MPI_DIST_GRAPH_CREATE_ADJACENT` as well (for both `sources` and `destinations`) on page 296 (298 in draft).
2. The following should be added at line 34 of page 297 (page 299 in draft) [it was added at line 32 of page 299 (35 of 301 in draft) instead, it's required in both]:
> MPI_PROC_NULL is a valid rank value for \mpiarg{sources} or for \mpiarg{destinations}.
3. Change-Log references wrong section and pages for these changes (§ 6.4.2 on page 239 instead of § 7.5.3 on page 296 and § 7.5.4 on pages 298, 299, 300, and 301)
4. Slight typo on page 300—should be "rank", not "rnak" :)

**Comment by dholmes-epcc-ed-ac-uk** [comment]
@omor1 thanks for the careful review 👍 

Here's the fixed version:
[mpi-report-ticket87.pdf](https://github.com/mpi-forum/mpi-issues/files/2030560/mpi-report-ticket87.pdf)

I expect that this errata change will be passed by the MPI Forum before issues #78, #84, and #93; the authors of those issues will be responsible for propagation of this change to new their functions.

**Comment by dholmes-epcc-ed-ac-uk** [comment]
@jdinan (21 hours ago, on the pull request, non-public - hence copied here)

> As long as we're clarifying this, it might be helpful to further clarify "MPI_PROC_NULL is valid ..., +and indicates that ...+" to also ensure the meaning of MPI_PROC_NULL is unambiguous.

@dholmes-epcc-ed-ac-uk just now

When explaining, we should cover several points:

* communication with null neighbours completes normally, like with point-to-point to/from null processes or off the edges of a cartesian topology [the easy tee shot]
* Graph and dist graph enquiry functions will include null neighbours just like non-null neighbours [the edge of the fairway]
* buffers/counts/datatypes/displacements/etc for null neighbours must be included for neighbourhood collectives - but not for other collectives [definitely into the rough now]

Alternatively, the graph and dist graph enquiry functions should remove all MPI_PROC_NULL values and give back only non-null neighbours (like a cartesian topology). That means getting back from a query function something different than what went into the creation function. In @omor1's example (issue mpi-forum/mpi-issues#87 (comment), comment 13th Mar, binary tree), the user must keep their neighbours input array, perform MPI_GROUP_TRANSLATE_RANKS to get ranks in the new communicator (including MPI_PROC_NULL->MPI_PROC_NULL mappings), and avoid usage of the topology query functions. However, "the sequence of neighbours is defined as the sequence returned by " (MPI-3.1, section 7.6, page 314) [looks like we've found a water hazard]

None of this is trivial, so I agree that some additional explanation is required here.

**Comment by dholmes-epcc-ed-ac-uk** [comment]
@omor1 I agree completely - my points in the list (I hope) follow the already-unambiguous text in §7.6 (without referencing it directly; thanks for finding the quotes) but show that *if* we start explaining (again) then the resulting text must be long and complex. Normal collectives are defined in terms of ```n``` point-to-point operations, where ```n``` is implied to be the result of MPI_COMM_SIZE, i.e. the number of real MPI processes in the communicator (e.g. MPI_GATHER, MPI-3.1, §5.5, p150, line 11, "as if the n processes in the group"). The "alternative" tries to make neighbourhood collectives like normal collectives but is, IMHO, untenable.

@jdinan Perhaps a cross-reference to §7.6 is sufficient? Such as "Section 7.6 describes how including MPI_PROC_NULL affects neighborhood collective operations."

@hjelmn This discussion about including MPI_PROC_NULL nodes (and duplicate nodes?) seems to indicate that the number of nodes in a distributed graph topology can be arbitrarily larger than the number of MPI processes in the communicator, which has implications for #89 (or, at least, for any future proposal to add MPI_DIST_GRAPH_MAP).

Given (MPI-3.1, §7.5.3, p294, line 35-36, re: MPI_GRAPH_CREATE):
> The call is erroneous if it specifies a graph that is larger than the group size of the input communicator.

and (MPI-3.1, §7.5.3, p295, line 45-46, re: MPI_GRAPH_CREATE):
> For a graph structure the number of nodes is equal to the number of processes in the group. Therefore, the number of nodes does not have to be stored explicitly.

it would seem that including MPI_PROC_NULL as a node in a non-distributed graph implies that one of the real MPI processes must be omitted as a node in that non-distributed graph.

The definitions for the distributed graph functions don't (I think) have this restriction.

**Comment by omor1** [comment]
@dholmes-epcc-ed-ac-uk @hjelmn counting `MPI_PROC_NULL` as a process in a communicator doesn't make sense. It isn't a 'real' process, and so poses a problem for collectives (MPI-3.1, §5.2.1, line 31-32):
>All processes in the group identified by the intracommunicator must call the collective routine.

`MPI_PROC_NULL` can't call anything, so it can't be a process in the group identified by an intracommunicator (whether it has a virtual process topology or otherwise).

Also note that while non-distributed graphs explicitly state the number of nodes in the resulting graph in the constructor, the distributed graph constructors do not (MPI-3.1, §7.5.4, p297 line 36-37 and p299 line 34-35):
>The number of processes in `comm_dist_graph` is identical to the number of processes in `comm_old`.

**Comment by schulzm** [comment]
The forum decided that should be handled as a full ticket and should be considered after a further investigation on implementation impact (having to check for NULL could lead to problems in optimized implementations). Either way, the forum decided that it should be clarified whether MPI_PROC_NULL is allowed or not

**Comment by wesbland** [comment]
@rlgraham32 will be able to more accurately represent his argument, but I think the main points were something like this:

Optimized implementations of neighborhood collectives (whether software or some future hardware) would not only have to check for `MPI_PROC_NULL` for the sake of not sending messages, but they would also have to shuffle buffers around to make sure the input and output buffers go to the correct places. Doing all of this extra shuffling is the concern as it may have a much greater performance impact than a single branch.

If an implementation is purely relying on point-to-point communication to implement neighborhood collectives (as all implementations we know of right now do), this has very little performance impact as you say because we _only_ have to check for `MPI_PROC_NULL`, which we already do during all of those communication calls.

**Comment by rlgraham32** [comment]
After thinking it over, I withdraw my concern.  Since proc null is not in the range of the communicator's ranks, and the implementation can store a version of the graph without the proc null's, an implementation can access only data that needs to be send/received w/o any special logic for handling the proc null case.

**Comment by wesbland** [comment]
Is that true? If you remove the `MPI_PROC_NULL`s from your list of neighbors, but the user's input to the collective will still include buffers (or perhaps `NULL` pointers) for those `MPI_PROC_NULL` MPI processes, don't you need to do some adjustment before transmitting or receiving buffers.

For instance, if these are your neighbors:

```
0 | 1, 2
1 | NULL, 0
2 | 0, 3
3 | NULL, 2
```

Your buffers on ranks 1 and 3 will include extra entries that you'll need to ignore.

**Comment by rlgraham32** [comment]
I looked at the neighborhood alltoall and alltoallv, and a single pointer is used for both source and destination buffers.  Data size and offsets, respectively, are used to get the address within the buffer, so there is really no portable way for at least the alltoall to specify an address for proc null.  I may have missed something ...

**Comment by wesbland** [comment]
Ok. That's probably fine then. I just wanted to make sure we weren't requiring a bunch of new null checks for every operation.

**Comment by omor1** [comment]
> the implementation can store a version of the graph without the proc nulls

Do you mean in addition to the version with them? They must be recorded somewhere, for the purpose of retrieving all neighbors with `MPI_DIST_GRAPH_NEIGHBORS_COUNT` and `MPI_DIST_GRAPH_NEIGHBORS` (MPI-3.1, §7.5.5, pp. 309, lines 37–41):
>The number of edges into and out of the process returned by `MPI_DIST_GRAPH_NEIGHBORS_COUNT` are the total number of such edges given in the call to `MPI_DIST_GRAPH_CREATE_ADJACENT` or `MPI_DIST_GRAPH_CREATE` (potentially by processes other than the calling process in the case of `MPI_DIST_GRAPH_CREATE`).

For this proposal, the input and/or output buffers could have blocks that are not modified by the neighborhood collectives, since the corresponding neighbor is `MPI_PROC_NULL`. These neighbors can't just be ignored; that bypasses the whole point of the proposal. In any case, this particular behavior is already mandated by the standard for cartesian topologies with non-periodic dimensions.

Did I misunderstand what you meant @rlgraham32? My impression of what you said was that an implementation could accept `MPI_PROC_NULL` neighbors, but then just ignore them.

**Comment by rlgraham32** [comment]
the version reference in my comment, just means an implementation copy of the graph - implementation private data.

I think you are right with your comment on proc null - I guess I was assuming that destinations were ordered such that the proc null in the graph would be first or last in the list with the example of the Cartesian shift, which I guess does not need to be the case.

So, one does need to check each neighbor to see if the data really needs to be sent or not - whether this is an mpi send routine, or some other logic, when MPI semantics are not relied on for the internal collective implementation logic.

As for Wesley's concern (and mine) about conditional logic, we can replace conditional logic by adding another vector internal to the implementation with offsets - another memory reference.

**Comment by omor1** [comment]
I expect null neighbors to be uncommon (seeing as this hasn't been brought up as an issue before now), so optimizing for the more common case where neighbors actually exist is probably better. In that case, the CPU branch predictors (which should be able to predict the branches with high accuracy, since no neighbor is `MPI_PROC_NULL`) might get better performance with conditional logic than having to dereference memory—but that's something that would require testing & optimization.

**Comment by bosilca** [comment]
Would this force implementations to support 2 versions, one optimized for the case without PROC_NULL and one generic (using more memory as suggested by @rlgraham32) ?

**Comment by dholmes-epcc-ed-ac-uk** [comment]
Is there still interest in pursuing this proposal?

It is marked as an errata (a little ambitious, perhaps), which means it has one chance to get into MPI-4.0 - specifically, it must be formally read at the next face-to-face meeting.

For the associated PR, I must choose between:
1) leave base branch as mpi-3.x, close the PR - target: unmerged
2) leave base branch as mpi-3.x, update the PR urgently, read at next meeting - target: errata for mpi-3.x
3) update base branch to mpi-4.x, update the PR eventually, read whenever - target: change for mpi-4.x

**Comment by omor1** [comment]
@dholmes-epcc-ed-ac-uk actually I [emailed the mpi-comments mailing list](https://lists.mpi-forum.org/pipermail/mpi-comments/2019-May/000277.html) a couple days ago about a related (and extended) version of this.

The gist is that 'rank' arguments (source, destination, root, target, etc) are variously described as either 'integer' or 'non-negative integer'. There's no real reason I can see for this. Additionally, it's unclear whether `MPI_PROC_NULL` is meant to be useable only in point-to-point communication or more generally wherever a 'rank' argument is allowed (e.g. one can imagine calling `MPI_PUT` with a null process as the target).

The proposal would be to add 'rank integer' as a semantic term with a possible value in the range [0, N) for a group of size N. If more expansive null process support is desired, then the definition of 'rank integer' can be extended to also include `MPI_PROC_NULL`; otherwise, this can be specified only for those functions in the point-to-point communication chapter. Receive calls will also need to specify that they can take `MPI_ANY_SOURCE` (which they already do).

I wanted to pass it through mpi-comments to get feedback before writing and submitting a proposal, but I guess no one actually uses mailing lists nowadays. Should I submit it on [mpi-forum/mpi-issues](/mpi-forum/mpi-issues), or to [mpiwg-semantic-terms/semantic-terms-issues](/mpiwg-semantic-terms/semantic-terms-issues)? It would touch large parts of the standard and the maintainers of the different chapters may want input. (I'd expect little opposition to clarifying 'integer' versus 'non-negative integer', but whether `MPI_PROC_NULL` should be allowed is evidently much more controversial.)

In any case, that proposal would supersede this one I think, since it's more general. Actually, it should probably be split in two for the two different portions...

**Comment by dholmes-epcc-ed-ac-uk** [comment]
@omor1 this proposal/issue could be expanded in scope to encompass the broader problems with inconsistent descriptions of rank in the standard. Closing this issue and starting another loses the history of comments, which dooms us to repeat them on the next issue.

Re-reading the comments on this issue, there is a clear need to make some clarifying text changes for this problem - so option (1) should be dropped.

The comment from @schulzm (https://github.com/mpi-forum/mpi-issues/issues/87#issuecomment-396747260) reminds me that the Forum decided this should be handled as a full change (2 votes), rather than errata (1 vote). In which case, this issue has probably already missed the MPI-4.0 boat - so option (2) should be dropped.

I think we are left with option (3) only.

**Comment by omor1** [comment]
@dholmes-epcc-ed-ac-uk right I agree, I think this is more complex than an errata and there's a clear need for clarification, so option (3) is left. I'l rewrite the proposal text and update the labels. Is there worth in keeping the current proposal text around for archival purposes?

**Comment by dholmes-epcc-ed-ac-uk** [comment]
@omor1 I think we could update this issue to encompass the wider context that you propose to tackle, rather than creating a new issue that would be disassociated from this conversation.

Alternatively, this issue (and its PR) can be closed and referenced from the new ones when they are created (with a couple of appropriate sentences of explanation).

**Comment by omor1** [comment]
I think that I'll create a _new_ issue for the 'integer'/'unsigned integer' issue and update this one to expand where `MPI_PROC_NULL` should/shouldn't be permitted. The problems are of course related, but mostly orthogonal.

**Comment by omor1** [comment]
Alright, I've updated this proposal. It now allows for clarification in either direction; i.e. either prohibit `MPI_PROC_NULL` from everywhere other than P2P calls or all it in all calls with 'rank' parameters. Or something in between, I suppose, as long as that is strictly clarified, though I'd like to see a strong rationale for that.

**Comment by dholmes-epcc-ed-ac-uk** [comment]
@omor1 is there still interest in pushing this issue forward? We are now running out of time for the mpi-4.0 release, just as we ran out of time for the mpi-3.1 release.

**Comment by devreal** [comment]
MPI 4 in §12.3 states that 

> `MPI_PROC_NULL` is a valid target rank in all MPI RMA communication calls.

It is also explicitly mentioned in `MPI_Win_shared_query` as a valid rank.

I consider this solved for the RMA chapter and will remove the label.

**Comment by Wee-Free-Scot** [comment]
@omor1 (and/or others) is there still interest in pushing this issue forward? We are now running out of time for the mpi-4.1 release, just as we ran out of time for the mpi-3.1 and mpi-4.0 releases.

I see four options:
1. We leave the text ambiguous and underspecified (this option has worked for decades, why change course now?)
2. We write "Use of MPI_PROC_NULL in graph or dist-graph topologies is erroneous." (bad)
3. We write "Use of MPI_PROC_NULL in graph or dist-graph topologies results in implementation-defined behavior." (ugly)
4. We write "Use of MPI_PROC_NULL in graph or dist-graph topologies is permitted and means ..." (good, but we have no consensus on what text finishes the sentence correctly.)
